dxr/plugins/xpidl/visitor.py

```diff
@@ -0,0 +1,252 @@
+            'icon': 'jump'
+        }
+
+    def visit_interface(self, interface):
+        # Yield refs for the members, methods, etc. of an interface (and the interface itself).
+        start = start_pos(interface.name, interface.location)
+        if interface.base:
+            # The interface that this one extends.
+            base_start = start_pos(interface.base, interface.location)
+            self.yield_ref(base_start, base_start + len(interface.base), [
+                self.filtered_search_menu('type-decl', interface.base, icon='type'),
+                self.subclass_search_menu(interface.base)
+            ])
+            self.yield_name_needle('derived', interface.base, interface.location)
+
+        self.yield_ref(start, start + len(interface.name), [
+            self.subclass_search_menu(interface.name),
+            self.generated_menu(interface)
+        ])
+        self.yield_name_needle('type_decl', interface.name, interface.location)
+
+        for member in interface.members:
+            if member.kind == 'const' or member.kind == 'attribute':
+                start = start_pos(member.name, member.location)
+                self.yield_ref(start, start + len(member.name), [
+                    self.filtered_search_menu('var-decl', member.name, icon='field'),
+                    self.filtered_search_menu('var', member.name, html='Find definition',
+                                              title='Search for definitions', icon='field'),
+                ])
+                self.yield_name_needle('var_decl', member.name, member.location)
+
+            elif member.kind == 'method':
+                member.location.resolve()
+                start = member.location._lexpos
+                self.yield_ref(start, start + len(member.name), [
+                    self.filtered_search_menu('function-decl', member.name, icon='method'),
+                    self.filtered_search_menu('function', member.name, 'Find implementations',
+                                              'Search for implementations of this method',
+                                              'method')
+                ])
+                lineno = self.check_lineno(member.name, member.location._lineno)
+                if lineno is not None:
+                    self.yield_needle('function_decl', {'name': member.name},
+                                      Extent(Position(lineno + 1, member.location._colno),
+                                             Position(lineno + 1,
+                                                      member.location._colno + len(member.name))))
+
+    def visit_include(self, item):
+        filename = item.filename
+        start = start_pos(filename, item.location)
+        # Remark: it would be nice if the Include class held on to this data after it does the
+        # same work.
+        # We know that the parser resolved the include in the same way, so we assume we can
```

Comment:
Yep, this is a likely place to optimize in the future, if profiles bear it out. It's a lot of repeated IO.